### ISSUE TITLE
[css-scrollbars][css-scrollbars-1] Using scrollbar-color to tint the scrollbar

### ISSUE THREAD

**Issue by felipeerias** [w3c/csswg-drafts#5651]
The property [scrollbar-color from CSS Scrollbars](https://drafts.csswg.org/css-scrollbars/#scrollbar-color) allows the user to pass two colors, to be used by the thumb and the track of the scrollbar.

So, `scrollbar-color: indigo lavender` with produce a scrollbar where the thumb is `#4B0082` and the track is`#E6E6FA`.

The spec also advises authors to _"ensure the specified colors have enough contrast between them"_.

I don't know if this has been discussed previously, but what should happen if only one color was provided to `scrollbar-color`?

```
scrollbar-color: orange;
```
```
scrollbar-color: darkgreen;
```

My proposal is that when one color is provided to `scrollbar-color`, it would be used by the UA to automatically define the color scheme for the scrollbar (to "tint" it, so to speak).

This would allow users to easily style the scrollbars on their pages while the UA ensures adequate accessibility by making sure that the specific colors have enough contrast between them.

--

One starting point to approach this, just to get the ball rolling, could be to use the provided color for the thumb and calculate the track color by keeping the same hue and changing lightness/saturation automatically until the desired contrast is found.

In the first example, the thumb would be `#FFA500` and the track would perhaps be `#362B17` (using this [contrast checker](https://colourcontrast.cc/362b17/ffa500) as reference).

In the second, the thumb would be `#006400` and the track would perhaps be `#E5ffE5` ([reference](https://colourcontrast.cc/e5ffe5/006400)).

This approach might not work in all cases, for instance with the `crimson` color `#DC143C`. For those cases, maybe the thumb color would need to be changed slightly to increase the contrast with the background. For example, `scrollbar-color: crimson;` might produce a `#9e0e2b`thumb and a `#fae9ed`track ([reference](https://colourcontrast.cc/fae9ed/9e0e2b)).

**Comment by felipeerias** [comment]
One starting point to approach this, just to get the ball rolling, could be to use the provided color for the thumb and calculate the track color by keeping the same hue and changing lightness/saturation automatically until the desired contrast is found.

In the first example above, the thumb would be `#FFA500` and the track would perhaps be `#362B17` (using this [contrast checker](https://colourcontrast.cc/362b17/ffa500) as reference).

In the second, the thumb would be `#006400` and the track would perhaps be `#E5ffE5` ([reference](https://colourcontrast.cc/e5ffe5/006400)).

This approach might not work in all cases, for instance with the `crimson` color `#DC143C`. For those cases, maybe the thumb color would need to be changed slightly to increase the contrast with the background. For example, `scrollbar-color: crimson;` might produce a `#9e0e2b`thumb and a `#fae9ed`track ([reference](https://colourcontrast.cc/fae9ed/9e0e2b)).

**Comment by frivoal** [comment]
This is separate, yet quite strongly related to the issues about accent color: https://github.com/w3c/csswg-drafts/issues/5187 / https://github.com/w3c/csswg-drafts/issues/5480 / https://github.com/w3c/csswg-drafts/issues/5544 / https://github.com/w3c/csswg-drafts/issues/5577

**Comment by LeaVerou** [comment]
> I don't know if this has been discussed previously, but what should happen if only one color was provided to `scrollbar-color`?

What [we recommended](https://github.com/w3ctag/design-reviews/issues/563#issuecomment-767375070) was slightly different: that each of the two color values also accepts the value `auto`, which produces a color generated by the UA. Then a single `<color>` value can just be a shorter form of `<color> auto` or `auto <color>`, whichever is more useful. 

> One starting point to approach this, just to get the ball rolling, could be to use the provided color for the thumb and calculate the track color by keeping the same hue and changing lightness/saturation automatically until the desired contrast is found.

On a high level, yes. However, as with most things color, this is not as simple and straightforward as it sounds:

- Using HSL lightness leads to unpredictable results, especially when the operation moves toward or crosses the 50% L boundary. E.g. increasing the lightness of `hsl(20, 100%, 30%)` (brown) by 50% produces `hsl(20, 100%, 45%)` (orange). Mixing with white or black, even if the interpolation is done in HSL, can produce better results (e.g. `hsl(20, 50%, 65%)`, a light brown).
- Regardless of the lightening/darkening algorithm, if the interpolation is done in a non perceptually uniform space like sRGB and friends (HSL, HWB etc) then we cannot define it as a X% adjustment, as it will produce completely different levels of perceptual difference. For example, [here](https://dabblet.com/gist/fd1f9b27342c28d4dfcc20263bf4ee23) is the same lightness adjustment, on colors of the same HSL lightness (50% to 70%):
![image](https://user-images.githubusercontent.com/175836/107970930-c7e2ee80-6f7f-11eb-8030-66ffbadc5287.png)
As you can see, the top pair is way more similar (ΔE2000 = 4.9), whereas the bottom pair is substantially different colors (ΔΕ2000 = 23.9). This can be solved by either interpolating in a perceptually uniform space (e.g. Lab or LCH), or by iterating and measuring the DeltaE between the foreground and background color, instead of having a naïve algorithm that performs the same percentage adjustment each time.
- The high level algorithm itself would need to be different in light and dark mode: in light mode the track is lighter than the thumb, and in dark mode, the track is darker than the thumb.
- What to do in edge cases? E.g. if the thumb is specified as white or black? I suppose it's better to darken instead of lighten and vice versa than to show an all-white or all-black unusable scrollbar.

**Comment by felipeerias** [comment]
Thank you very much for your detailed feedback, @LeaVerou 

I thought that there could be a way to define what would happen if only one color was provided to `scrollbar-color`, but it seems that this syntax is too ambiguous to provide a good outcome in all cases.

I like your proposal to use a keyword to indicate that a part's color should be automatically calculated by the UA. Do you want to open a separate issue for that?

**Comment by frivoal** [comment]
I don't think we should get into the details of specifying exactly how the browser computes the contrasting color. It should just do so, in a way that ensure good contrast, and how to do that is up to the UA.

Also, do we want: `auto | <color> {1,2}` or `[auto | <color>]{1,2}`? With the latter, when providing a single color, you can decide if it's for the track or the thumb, while with the first one, you just give a tint and let the UA worry about where it goes.
